Raft test topology part 1

alecco opened the pull request:
These are the first commits out of #10815 and is itself based on #11193 .

It starts by moving pytest logic out of the common `test/conftest.py` and into `test/topology/conftest.py`, including removing the async support as it's not used anywhere else.

There's a fix of a bug of leaving tables in `RandomTables.tables` after dropping all of them.

Keyspace creation is moved out of `conftest.py` into `RandomTables` as it makes more sense and this way topology tests avoid all the workarounds for old version (topology needs ScyllaDB 5+ for Raft, anyway).

And a minor fix.

kbr- commented on test/pylib/random_tables.py:
```diff
@@ -240,6 +240,8 @@ def __init__(self, test_name: str, cql: CassandraSession, keyspace: str):
         self.keyspace = keyspace
```
Comments regarding the commit message:

1. "creation and drop should happen only once per cluster and no test should be running changes (either not started or finished)"

I had to think for a few minutes to interpret this.

Is this the corrent understanding: "creation and drop **of the keyspace** should happen only once per cluster and no test should be running changes (either not started or finished) **while the keyspace is being created or dropped**"?

2. All topology tests are for Scylla with Raft. So don't use the Cassandra
this_dc workaround as it's unnecessary for Scylla.

Nowhere in this patch we can find the string `this_dc`. If you didn't mention it in the commit message, I wouldn't even know that there's a problem with Cassandra and we needed some workaround for Cassandra at some point (why should I even care?)

And now I have to ask: **what** is the `this_dc` workaround? Sorry.

kbr- reviewed (approved):
Just some questions, if my understanding of everything is right, LGTM.

But I still have a problem with https://github.com/scylladb/scylladb/pull/11193 - I posted a comment there.

alecco commented on test/topology/conftest.py:
```diff
@@ -6,10 +6,75 @@
+    return _wrap_future(self.execute_async(*args, **kwargs))
+
+
+Session.run_async = run_async
```
> This monkey-patching applies only to tests which explicitly import test/topology/conftest.py, right?
...
> Unless run_async is not an existing member in Session, we're adding a completely new member? Then it doesn't matter.

Yes, nothing else is using it.

> Doesn't pytest use this file internally somehow? I'm worried if pytest won't accidentally apply this patch to other test modules as well due to some pytest magic.

No, this is the cassandra driver. And pytest processes are executed independently for each pytest file. There is no way for this to affect anything outside topology test directory. It was before but no longer. This series reverts previous async changes to top level conftest wich _could_ affect other tests but it _did not_.

> If it would, maybe a way to protect ourselves from that is to apply the patch in the cql fixture.

If monkey patching is not wanted (even though it's used elsewhere), I would rather subclass `cassandra.Session`. Let me know.

kbr- replied:
> If monkey patching is not wanted (even though it's used elsewhere), I would rather subclass cassandra.Session than keep adding fixture hacks. Let me know.

No, I don't think doing more work here would be valuable, let's leave it as it is.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1665/

alecco commented:
In latest push I added `pytest_asyncio` mode auto. Removed leftover `pytest.mark.asyncio`.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1672/

kbr- commented:
> In latest push I added pytest_asyncio mode auto. Removed leftover pytest.mark.asyncio.

Can you discuss this with @kostja ? This seems contradictory to @kostja 's earlier change: https://github.com/scylladb/scylladb/pull/11155

alecco commented:
> > In latest push I added pytest_asyncio mode auto. Removed leftover pytest.mark.asyncio.
> 
> Can you discuss this with @kostja ? This seems contradictory to @kostja 's earlier change: #11155

I discussed with Kostja. His pull request was a workaround for not being able to have `auto` mode. This change removes the need for his pull request.

kostja commented:
> > In latest push I added pytest_asyncio mode auto. Removed leftover pytest.mark.asyncio.
> 
> Can you discuss this with @kostja ? This seems contradictory to @kostja 's earlier change: #11155

yeah, it's fine,I already tested alejo's branch locally here this morning and it passes all checks. The bigger issue with mode=auto is that you don't have to explicitly mark fixtures/tests as async, which is evil, because a function can sneak in which will just blow up (very confusing error output) once you upgrade to a newer pytest_asyncio. But as I wrote elsewhere, this is going to be a problem when we upgrade our build toolchain to pytest-asyncio 0.19.

Anyhow, I am more than willing to rebase my pull request on top of this one to let it go first. This work is more important.

kbr- commented:
> Anyhow, I am more than willing to rebase my pull request on top of this one to let it go first. This work is more important.

Well, if we're eventually planning to go with the `strict` version, then I wonder why update this PR to include the additional commit at all. Seems like unnecessary work. The previous version was fine.

Anyway, @alecco I just queued the https://github.com/scylladb/scylladb/pull/11193. You can rebase this on top of 97f0e11c3a9042385c3a209994912b12577d1112 so CI can start running, and once next build passes, we'll be able to queue this PR as well (probably tomorrow).

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1682/

kbr- commented:
@alecco you didn't rebase on top of https://github.com/scylladb/scylladb/commit/97f0e11c3a9042385c3a209994912b12577d1112 as I proposed (this commit was merged to `master`), the PR contains the commit ` test.py: handle properly pytest ouput file for CQL tests ` with a different hash (446d9ad699401f14c3dc5509d8678670832d35a1) so we have a conflict. So another rebase is needed... This time you can rebase directly on `master`.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1692/
